scr/src/main/java/org/apache/felix/scr/impl/inject/field/FieldHandler.java

```diff
@@ -171,9 +171,9 @@ private boolean initField(final Object componentInstance,
         return objects;
     }
 
-    private MethodResult updateField(final METHOD_TYPE mType,
-                                     final Object componentInstance,
-                                     final BindParameters bp)
+    private synchronized MethodResult updateField(final METHOD_TYPE mType,
```

Comment:
I think this synchronization is too broad.  In particular I am concerned about synchronization when making the calls to `org.apache.felix.scr.impl.inject.ValueUtils.getValue(String, ValueType, Class<?>, ScrComponentContext, RefPair<?, ?>)` which ultimately can set off a chain of getService calls that may trigger activating dependency components.

A more safe fix is to use a synchronized map and synchronize on the map around the bit of code that needs it:

```java
                Map<RefPair<?, ?>, Object> boundValues = bp.getComponentContext().getBoundValues(metadata.getName());
                synchronized (boundValues)
                {
                    if ( this.metadata.isOptional() && !this.metadata.isStatic() )
                    {
                        // we only reset if it was previously set with this value
                        if ( boundValues.size() == 1 )
                        {
                            this.setFieldValue(componentInstance, null);
                        }
                    }
                    boundValues.remove(refPair);
                }
```

Then I would just do `Collections.synchronizedMap(new TreeMap<>(...` when creating the map for bound values.

Reply:
I saw it coming 😉 Yes, you are right. The `synchronized` block was a bit broader in this respect. We would be better off with synchronizing only the part that requires it. I will do the needful.

Reply:
Applied your proposal